### PR TITLE
Fetch and filter records for Alma MARC harvest

### PR DESCRIPTION
**Pull request by ghukill**
### Purpose and background context
This PR introduces the ability for GeoHarvester to harvest Alma MARC records.  This first update only includes the ability to fetch and filter geospatial MARC records from S3.  The files it is targeting are MARC XML files deposited by the TIMDEX Alma exports.

By introducing a new `MITAlmaHarvester` harvester, this paves the way for a new `gisalma` TIMDEX source that would sit alongside of `gismit` and `gisogm`.

It has been observed that this is not the most efficient way to harvest a subset of Alma MARC records that meet our geospatial requirements, but it is a path with minimal (if any) infrastructure changes at the moment.  If and when other mechanisms to produce a smaller subset of Alma MARC records emerge, this harvester could be updated to point at a different location, but the mechanics should remain largely the same, potentially with some filtering still happening at this application level.

PR review notes and areas of focus:
- the python library [marcalyx](https://pypi.org/project/marcalyx/) was selected to parse MARC XML records
  - because it parses from an XML string, well suited for the lazy evaluation approach of scanning through records; vs `pymarc` which is more aligned with reading directly from a file-object
- the `MARC` `SourceRecord` class is **stubbed** at this time to support the full `get_source_records()` functionality 
  -  no field methods are defined yet, which is slated for [GDT-227](https://mitlibraries.atlassian.net/browse/GDT-227)
- the methods `full_harvest_get_source_records()` and `incremental_harvest_get_source_records()` both point to the same method `_get_source_records()`
  - the addition of the CLI command may tweak this slightly, but unlike MIT and OGM harvesters, the mechanics are virtually the same, just driven entirely through `harvest-type` and `from-date / until-date` properties

### How can a reviewer manually see the effects of these changes?

As there is no CLI command yet for `gisalma` harvests, a python shell is required:

```python
import logging
from harvester.config import configure_logger
from harvester.harvest.alma import MITAlmaHarvester

root_logger = logging.getLogger()
configure_logger(root_logger, True)

# init harvester that will read from S3
harvester = MITAlmaHarvester(
    input_files="s3://timdex-extract-dev-222053980223/alma",
    harvest_type="full",
    from_date="2023-01-13",
    until_date="2023-01-14",
)

# get iterator of records
records = harvester.full_harvest_get_source_records()

# scan until first record that matches geospatial MARC filtering
record = next(records)

# assert both raw XML stored and parsed MARCRecord on record.source_record
assert record.source_record.data.startswith(b"<record><leader>02723nem  2200589Ii 4500</leader>")
assert record.source_record.marc.titleStatement().value.startswith("Gravity-Bouguer anomalies, Ottawa Islands, Northwest Territories")
```

To also get a sense of how the harvester scans through many MARC records but filters to a subset:
```python
import itertools

# retrieve 100 records demonstrating scan through 50k+ records
# (if performed in same python console with logger set to DEBUG above, will see logging outputs)
record_subset = list(itertools.islice(records, 100))
```

### Includes new or updated dependencies?
YES

### Changes expectations for external applications?
NO

### What are the relevant tickets?
- https://mitlibraries.atlassian.net/browse/GDT-226

### Developer
- [ ] All new ENV is documented in README
- [ ] All new ENV has been added to staging and production environments
- [ ] All related Jira tickets are linked in commit message(s)
- [ ] Stakeholder approval has been confirmed (or is not needed)

### Code Reviewer(s)
- [ ] The commit message is clear and follows our guidelines (not just this PR message)
- [ ] There are appropriate tests covering any new functionality
- [ ] The provided documentation is sufficient for understanding any new functionality introduced
- [ ] Any manual tests have been performed and verified
- [ ] New dependencies are appropriate or there were no changes

[GDT-227]: https://mitlibraries.atlassian.net/browse/GDT-227?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by ehanson8 — COMMENTED** [review]
Looking good, a few notes

File: harvester/harvest/alma.py [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+
+    def full_harvest_get_source_records(self) -> Iterator[Record]:
+        """Identify files for harvest by parsing MARC records from FULL Alma exports."""
+        CONFIG.check_required_env_vars()
+        yield from self._get_source_records()
+
+    def incremental_harvest_get_source_records(self) -> Iterator[Record]:
+        """Identify files for harvest by parsing MARC records from DAILY Alma exports."""
+        CONFIG.check_required_env_vars()
+        yield from self._get_source_records()
```
[the remark this answers is not in the document]
Comment (jonavellecuerdo) [reply harvester/harvest/alma.py:43]: Second this question!

File: harvester/harvest/alma.py [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+
+    def _get_source_records(self) -> Iterator[Record]:
```
Comment (jonavellecuerdo) [inline harvester/harvest/alma.py:45]: What is the difference between a full and incremental harvest for Alma?

File: Pipfile [hunk 3]
```diff
@@ -18,6 +18,7 @@ pygit2 = "*"
 shapely = "*"
 requests = "*"
 pyaml = "*"
+marcalyx = "*"
```
[the remark this answers is not in the document]
Comment (ghukill) [reply Pipfile:21]: I thought so too @ehanson8!  It hasn't had many updates in recent years...but neither has MARC, a 50 year old format.  So far, the ergonomics are great for parsing data, and it's quick.

File: harvester/harvest/alma.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (ghukill) [reply harvester/harvest/alma.py:43]: These methods are required by the base `Harvester` class.  Based on the `harvest-type` from the CLI command, the approporiate method is called from the extending `Harvester` class.  For MIT and OGM harvesters, they are different, but the Alma harvester is a bit unique in that the code paths are the same.  

I added a docstring comment to the shared `_get_source_records()`, but I don't think it'd hurt to add one here as well.

File: harvester/harvest/alma.py [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+        for filepath in self._list_xml_files():
+            with smart_open.open(filepath, "rb") as f:
+                context = etree.iterparse(f, events=("end",), tag="record")
+                for _event, elem in context:
+                    yield MARCRecord(elem)
+                    elem.clear()
+                    while elem.getprevious() is not None:
+                        del elem.getparent()[0]
+
+    def filter_geospatial_marc_records(
```
[the remark this answers is not in the document]
Comment (ghukill) [reply harvester/harvest/alma.py:122]: Great suggestion; done.

File: harvester/harvest/alma.py [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+        for marc_record in self.filter_geospatial_marc_records(all_marc_records):
+            identifier, source_record = self.create_source_record_from_marc_record(
+                marc_record
+            )
+            yield Record(
+                identifier=identifier,
+                source_record=source_record,
+            )
+
+    def _list_xml_files(self) -> Iterator[str]:
```
[the remark this answers is not in the document]
Comment (ghukill) [reply harvester/harvest/alma.py:57]: I felt the same, thanks for the nudge.  I see only benefits from breaking it up, will update.

File: harvester/harvest/alma.py [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+        Example filepath: alma-2024-03-01-daily-extracted-records-to-index_19.xml
+            - run_type=extracted
+            - run_date=2024-03-01
+        """
+        if self.input_files.startswith("s3://"):
+            filepaths = self._list_s3_xml_files()
+        else:
+            filepaths = self._list_local_xml_files()  # pragma: nocover
+
+        harvest_type_map = {"full": "full", "incremental": "daily"}
```
[the remark this answers is not in the document]
Comment (ghukill) [reply harvester/harvest/alma.py:69]: This is a translation I think we'll have to keep.  The GeoHarvester is baked with "full" and "incremental" as the names of harvest types, but we'll need to look for "full" or "daily" in the filepaths from S3.

File: harvester/harvest/alma.py (hunk 2 above)
Comment (ghukill) [reply harvester/harvest/alma.py:45]: At this time, the only difference is the `Harvester.harvest_type` property that is set.  When the flag is `full`, then we look for `full` in the S3 filepaths.  When flag is `incremental`, then we look for `daily` in the S3 filepath.

This harvester is a bit more streamlined vs MIT and OGM, as the files are in the same location, and have the same structure.  The only difference is identifying _which_ files to include.  Hence, the date filtering as well.

File: harvester/harvest/alma.py [hunk 7]
```diff
@@ -0,0 +1,192 @@
...
+        self, marc_records: Iterator[MARCRecord]
+    ) -> Iterator[MARCRecord]:
+        """Yield geospatial MARC records by filtering on defined criteria."""
+        for i, record in enumerate(marc_records):
+            if i % 10_000 == 0 and i > 0:  # pragma: nocover
+                message = f"{i} MARC records scanned for geospatial filtering"
+                logger.info(message)
+
+            # skip if leader doesn't have a/c/n/p
+            if record.leader[5] not in ("a", "c", "n", "p"):
```
Comment (jonavellecuerdo) [inline harvester/harvest/alma.py:132]: I understand that leader code "d" means "Record is deleted". It seems like this these records are skipped from processing. What does this mean in the context of the harvest? Wouldn't it inform us which files to delete (if we did end up copying files to the CDN/Public S3 buckets)?

File: harvester/harvest/alma.py [hunk 8]
```diff
@@ -0,0 +1,192 @@
...
+                continue
+
+            # skip if Genre/Form 655 does not contain "Maps."
+            if not any(
+                "Maps." in subfield.value
```
Comment (jonavellecuerdo) [inline harvester/harvest/alma.py:137]: Perhaps we should make compare the values in lowercase formats? For example, `alma-2023-12-31-daily-extracted-records-to-index_01.xml` contains the following records. 
image

It seems like _at least one_ entry will always contain "Maps." if it is a GIS record, but this change wouldn't hurt (I think)!

Maybe:

```
"maps" in subfield.value.rstrip(".").lower().split()
for form in record["655"]
for subfield in form.subfield("a")
```

File: harvester/harvest/alma.py [hunk 9]
```diff
@@ -0,0 +1,192 @@
...
+                for form in record["655"]
+                for subfield in form.subfield("a")
+            ):
+                continue
+
+            # skip if call number prefix not in list
+            if not any(
```
Comment (jonavellecuerdo) [inline harvester/harvest/alma.py:144]: Same comment re: letter casing.

File: harvester/harvest/alma.py [hunk 10]
```diff
@@ -0,0 +1,192 @@
...
+                subfield.value in ("MAP", "CDROM", "DVDROM")
+                for locations in record["949"]
+                for subfield in locations.subfield("k")
+            ):
+                continue
+
+            # skip if shelving location not in list
+            if not any(
+                subfield.value in ("MAPRM", "GIS")
```
Comment (jonavellecuerdo) [inline harvester/harvest/alma.py:153]: Same comment re: letter casing.

File: harvester/harvest/alma.py [hunk 11]
```diff
@@ -0,0 +1,192 @@
...
+                for locations in record["985"]
```
Comment (jonavellecuerdo) [inline harvester/harvest/alma.py:154]: [Non-blocking]: I continue to be curious about the schema from MIT-owned sources and wonder how we can gain access to the mappings. From my understanding, code 985 is a custom field, right?

File: harvester/harvest/alma.py (hunk 7 above)
Comment (jonavellecuerdo) [reply harvester/harvest/alma.py:132]: If records where `record.leader[5] == "d"` aren't being yielded, then lines 175-184 will never result in a `deleted` event, right?

**Review by jonavellecuerdo — CHANGES_REQUESTED** [review]
Nice work on this one! I agree that the diagrams helped ease review, and the code itself was also well-written and easy to understand. I may have 2 change requests:

* Consider letter casing in `MITAlmaHarvester.filter_geospatial_marc_records()`
* Yield records where `records.leader[5] == "d"`

File: harvester/harvest/alma.py [hunk 12]
```diff
@@ -0,0 +1,192 @@
...
+            if i % 10_000 == 0 and i > 0:  # pragma: nocover
+                message = f"{i} MARC records scanned for geospatial filtering"
+                logger.info(message)
+
+            # skip if leader doesn't have a/c/n/p
+            if record.leader[5] not in ("a", "c", "n", "p"):
+                continue
+
+            # skip if Genre/Form 655 does not contain "Maps."
+            if not any(
```
Comment (jonavellecuerdo) [inline harvester/harvest/alma.py:136]: Perhaps we should make compare the values in lowercase formats? For example, `alma-2023-12-31-daily-extracted-records-to-index_01.xml` contains the following records. 
image

It seems like _at least one_ entry will always contain "Maps." if it is a GIS record, but this change wouldn't hurt (I think)!

Maybe:

```
"maps" in subfield.value.rstrip(".").lower().split()
for form in record["655"]
for subfield in form.subfield("a")
```

File: harvester/harvest/alma.py (hunk 7 above)
Comment (ghukill) [reply harvester/harvest/alma.py:132]: Great catch @jonavellecuerdo!  Wil add.  That was porting legacy Slingshot, but it was not concerned with deleted records, while we are.

File: harvester/harvest/alma.py (hunk 11 above)
Comment (ghukill) [reply harvester/harvest/alma.py:154]: It's a great question, and I think related to @ehanson8's comment about stakeholder review and the complexity of this logic.

My vote would be to move forward with this logic as-is for now, knowing it still needs some stakeholder review.  Once we settle on the particular MARC fields and qualities that will filter to our desired records -- and maybe that _is_ this -- then I'd propose we add a standalone markdown doc that outlines the filtering criteria.

File: harvester/harvest/alma.py (hunk 9 above)
Comment (ghukill) [reply harvester/harvest/alma.py:144]: I believe that casing is well controlled in MARC records, particularly for these fields.  But as we tune the filtering, that could be a good thing to verify.

File: harvester/harvest/alma.py (hunk 10 above)
Comment (ghukill) [reply harvester/harvest/alma.py:153]: Same here.

File: harvester/harvest/alma.py (hunk 12 above)
Comment (ghukill) [reply harvester/harvest/alma.py:136]: For the time being, I'd prefer to keep it faithful to the original Slingshot application's logic.  This resulted in a known set of MARC records that satisfied stakeholders at the time.  

But these are good suggestions and worth looking into as we tune this filtering.

**Comment by ghukill** [comment]
@ehanson8, @jonavellecuerdo - a [new commit](https://github.com/MITLibraries/geo-harvester/pull/205/commits/411a2e4a6821ff7dc23f5b0e2fd085d967c47e42) has been added with revisions from the PR, and a review re-requested.

**Review by jonavellecuerdo — APPROVED** [review]

File: harvester/harvest/alma.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (ehanson8) [reply harvester/harvest/alma.py]: Got it, thank you for clarifying!

**Pull request merged**